src/charm_state.py

```diff
@@ -167,20 +213,54 @@ def from_charm(cls, charm: CharmBase) -> "State":
         Returns:
             Current state of the charm.
         """
+        prev_state = None
+        if CHARM_STATE_PATH.exists():
+            with open(CHARM_STATE_PATH, mode="r", encoding="utf-8") as state_file:
```

Comment:
Would `Path().read_text(encoding="utf-8")` be useful here?

Reply:
Use the `read_text` with `json.loads` to reduce nesting.